File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        st = []
        n = len(heights)
        for i in range(n+1):
            while st and (i==n or heights[st[-1]] > heights[i]):
                height = heights[st.pop()]
                width = i if not st else i - st[-1] - 1
                maxArea = max(maxArea, height * width)
            st.append(i)
        return maxArea

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        //Monotonic stock
        //two pointers
        int maxArea = 0;
        int n = heights.length;
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<=heights.length;i++){
            while(!st.isEmpty() && (i==n || heights[st.peek()]>=heights[i])){
                int height = heights[st.pop()];
                int width = st.isEmpty()?i : i - st.peek() - 1;
                maxArea = Math.max(maxArea, height*width);
            }
            st.add(i);
        }
        return maxArea;
    }
}